### ISSUE TITLE
Misleading errors msg with sparse precomputed matrix

### ISSUE THREAD

**Issue by thomasopsomer** [scikit-learn-contrib/hdbscan#253]
Hi,

When using hdscan with a sparse precomputed distance matrix, it may throw an error saying that the distance matrix has more than one component and that one should apply dbscan on each component.

However, this error / message can also happen in case where the distance matrix has only one component, leading to some frustration :).

Looking at the implementation, the error is thrown looking at the `mutual_reachability`. The way it is computed can indeed lead to a `mutual_reachability` having several component while the distance matrix has only one, because if the reachability is 'infinite' (no k'th neighbor) then the mutual reachability becomes also infinite and isn't stored in the `mutual_reachability` matrix (https://github.com/scikit-learn-contrib/hdbscan/blob/master/hdbscan/_hdbscan_reachability.pyx#L91)

So concerning the error maybe we could actually have 2 errors: a first one raised if the distance matrix have more than 1 component, and then a second one if there an issue with the `mutual_reachability` ?

Another question, why can't we set a max value in the `mutual reachability` matrix when the reachability in infinite ?

**Comment by lmcinnes** [comment]
I agree, and this has been on my TODO list to fix eventually. Unfortunately I haven't had much time to devote to hdbscan recently. If you want to submit a PR I would be happy to merge it.

**Comment by thomasopsomer** [comment]
No pb for the PR :)

Concerning setting a max value in the `mutual_reachability` when the reachability is infinite, do you think it's good idea ?
